contracts/src/Gateway.sol

```diff
@@ -274,12 +280,24 @@ contract Gateway is IGateway, IInitializable, IUpgradable {
             revert InvalidAgentExecutionPayload();
         }
 
-        bytes memory call = abi.encodeCall(AgentExecutor.execute, params.payload);
+        (AgentExecuteCommand command, bytes memory commandParams) =
+            abi.decode(params.payload, (AgentExecuteCommand, bytes));
+        if (command == AgentExecuteCommand.Transact) {
+            (address target, bytes memory payload,) = abi.decode(commandParams, (address, bytes, uint64));
+            SafeCallFilterStorage.Layout storage $ = SafeCallFilterStorage.layout();
+            bytes4 selector = bytes4(payload.slice(0, 4));
+            if ($.safeCalls[target][selector] != true) {
+                revert NoPermission();
+            }
+        }
```

Comment:
So this is just a temporary measure to release this feature safely to allowed chains?

Reply:
i worry that this safe call filter both here and in your past PRs is an anti-pattern that doesn't just limit the functionality of the bridge, but also leads us to bad architectural design decisions that will have wider impact on fees, execution/channels and our trust model overall

Reply:
Actually we can't allow transact to invoke arbitrary call, at least not including the `AgentExecutor::mintToken|burnToken` which is introduced in polkadot native assets.

More context in https://github.com/Snowfork/snowbridge/pull/1155#issuecomment-2026820489

I can probably change it to a blacklist, the whitelist is just follow the pattern every parachain should declare which calls are exposed through xcm.

Reply:
> The real thing we need here is a different Agent for transact than the Agent we use to lock funds in, maybe an agent derived from a pallet on parachain or user.

Yeah, I like this. The isolation with different Agent make more sense here. I'll make change accordingly.

Reply:
https://github.com/Snowfork/polkadot-sdk/pull/116/commits/f4b6dba04744f6c31a589fe7ccb83cdc3be886f5#diff-952be233381eef865196222dc9f95b0d88e54cdcec924e84d03c5622d2e59ec9 and https://github.com/Snowfork/snowbridge/pull/1145/commits/606e867b7badc6d356c8f4b56e6b81ee0eb27811

Reply:
>>The real thing we need here is a different Agent for transact than the Agent we use to lock funds in, maybe an agent derived from a pallet on parachain or user.

> Yeah, I like this. The isolation with different agents make more sense here. I'll make change accordingly.

Agree